File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    return dfs(root)[1]
};

var dfs = (root) =>{
    if (root === null) return [0, 0] // [hight, diameter]
    
    let left             = dfs(root.left)
    let right            = dfs(root.right)
    let longest_diameter = Math.max(left[1], right[1], left[0] + right[0])
        
    return [1 + Math.max(left[0], right[0]), longest_diameter]
}      

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        longest_diameter = [0]
        def dfs(root):
            if not root:
                return 0
        
            left                = dfs(root.left)
            right               = dfs(root.right)
            longest_diameter[0] = max(longest_diameter[0], left + right)
        
            return 1 + max(left, right)
        dfs(root)
        return longest_diameter[0]
        
        
        """def dfs(root):
            if not root:
                return [0, 0] # [hight, diameter]
        
            left             = dfs(root.left)
            right            = dfs(root.right)
            longest_diameter = max(left[1], right[1], left[0] + right[0])
        
            return [1 + max(left[0], right[0]), longest_diameter]
        
        return dfs(root)[1]"""